File: app/src/main/java/com/example/newsapp/viewModel/WikiNewsViewModel.java
package com.example.newsapp.viewModel;

import android.app.Application;
import android.content.Context;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.os.AsyncTask;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.example.newsapp.model.WikiNews;
import com.example.newsapp.utils.FetchWikiArticleTask;

import java.util.ArrayList;
import java.util.List;

public class WikiNewsViewModel extends AndroidViewModel {

    private boolean dataLoaded = false;
    private MutableLiveData<List<WikiNews>> wikiNewsLiveData = new MutableLiveData<>();

    public WikiNewsViewModel(@NonNull Application application) {
        super(application);
    }

    public LiveData<List<WikiNews>> getWikiNewsLiveData() {
        return wikiNewsLiveData;
    }

    public void loadData() {
        if (isInternetAvailable()) {
            new FetchDataTask().execute("Android_(operating_system)");
        } else {
            wikiNewsLiveData.setValue(new ArrayList<>());
        }
    }

    public boolean isInternetAvailable() {
        ConnectivityManager connectivityManager = (ConnectivityManager) getApplication().getSystemService(Context.CONNECTIVITY_SERVICE);
        if (connectivityManager != null) {
            NetworkInfo networkInfo = connectivityManager.getActiveNetworkInfo();
            return networkInfo != null && networkInfo.isConnected();
        }
        return false;
    }

    public boolean isDataLoaded() {
        return dataLoaded;
    }

    private class FetchDataTask extends AsyncTask<String, Void, String> {
        @Override
        protected String doInBackground(String... params) {
            // Perform the network operation in the background
            return FetchWikiArticleTask.fetchWikiArticles().toString();
        }

        @Override
        protected void onPostExecute(String htmlContent) {
            // Update the LiveData on the main thread
            ArrayList<WikiNews> wikiNewsList = new ArrayList<>();
            if (htmlContent != null) {
                WikiNews wikiNews = new WikiNews("Android OS", htmlContent, "https://en.wikipedia.org/wiki/Android_(operating_system)");
                wikiNewsList.add(wikiNews);
                dataLoaded = true;
            }
            wikiNewsLiveData.setValue(wikiNewsList);
        }
    }
}

File: app/src/main/java/com/example/newsapp/viewModel/WikiNewsViewModel.kt
package com.example.newsapp.viewModel

import android.app.Application
import android.content.Context
import android.net.ConnectivityManager
import android.os.AsyncTask
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.example.newsapp.model.WikiNews
import com.example.newsapp.utils.FetchWikiArticleTask

class WikiNewsViewModel(private val application: Application) : ViewModel() {
    var isDataLoaded: Boolean = false;
    val wikiNewsLiveData = MutableLiveData<List<WikiNews?>?>()

    fun getWikiNewsLiveData(): LiveData<List<WikiNews?>?> {
        return wikiNewsLiveData
    }

    fun loadData() {
        val isInternetAvailable = isInternetAvailable()

        if (isInternetAvailable) {
            FetchDataTask().execute()
        } else {
            wikiNewsLiveData.value = null
        }
    }

    fun isInternetAvailable(): Boolean {
        val connectivityManager =
            application.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        val networkInfo = connectivityManager.activeNetworkInfo
        return networkInfo != null && networkInfo.isConnected
    }

    private inner class FetchDataTask : AsyncTask<Void?, Void?, List<WikiNews?>?>() {


        override fun onPostExecute(wikiNewsList: List<WikiNews?>?) {
            // Update the LiveData on the main thread
            wikiNewsLiveData.value = wikiNewsList
            isDataLoaded = true
        }

        override fun doInBackground(vararg params: Void?): List<WikiNews?>? {
            TODO("Not yet implemented")
            return FetchWikiArticleTask.Companion.fetchWikiArticles()

        }
    }
}
